doc/smart-switch/smart-switch-database-architecture/smart-switch-database-design.md

```diff
@@ -0,0 +1,262 @@
+
+
+### Requirements
+
+- All databases, including those on both NPU and DPU, must be accessible through the GNMI server.
+- Each DPU database instance on the NPU is associated with a unique TCP port and domain Unix socket path.
+- All DPU database instances on the NPU will be bound to the IP address of the midplane bridge.
+- All database instances on the NPU share the same network namespace to facilitate seamless communication.
+- DPUs can access their respective overlay database instances using the IP of the midplane bridge and a pre-assigned unique TCP port.
+
+### Architecture Design
+
+![smart-switch-database-architecture](smart-switch-database-architecture.png)
+
+#### Database services
+
+In this section, the focus is on illustrating the maintenance of DPU overlay databases within the NPU. It's essential to note that the traditional database services of both NPU and DPU remain unchanged and do not necessitate further design modifications.
+
+The management of DPU overlay databases within the NPU is orchestrated through existing SONiC database services. The daemon, named "featured," retains the responsibility for initiating, terminating, enabling, and disabling the DPU overlay database services. This interaction is facilitated using the systemctl tool.
+
+To determine the DPU number, the "featured" daemon should leverage the platform API. However, for the sake of implementation simplicity, the DPU number is extracted directly from the platform_env.conf file firstly.
+
+``` shell
+cat /usr/share/sonic/device/$PLATFORM/platform_env.conf
+NUM_DPU=2
+```
+
+To align with the established multi-ASIC design in SONiC, a new field, `"has_per_dpu_scope": "True"``, is introduced in the database feature table within config_db.json. This field plays a crucial role in ensuring that each DPU database instance is initiated within a dedicated database container. This design approach maintains consistency with SONiC's existing architecture while accommodating the specific requirements of DPU overlay databases.
+
+``` json
+# config_db.json
+
+"database": {
+    "auto_restart": "always_enabled",
+    "delayed": "False",
+    "has_global_scope": "True",
+    "has_per_asic_scope": "True",
+    "has_per_dpu_scope": "True", # New field for DPU database service
+    "high_mem_alert": "disabled",
+    "state": "always_enabled",
+    "support_syslog_rate_limit": "true"
+},
+```
+
+Within the NPU, the management of DPU overlay databases involves specific configurations. Each DPU overlay database instance is bound to the IP address of the midplane bridge (169.254.200.254 by default). The TCP port assignment follows a predictable pattern, with each DPU ID associated with a unique port (6380 + DPU ID). Additionally, the Unix domain socket path is structured as /var/run/redisdpu{DPU_ID}.
```

Comment:
>/var/run/redisdpu{DPU_ID} [](http://example.com/codeflow?start=383&length=25)

This rule should be defined in runtime config file `database_global.json`.

Reply:
Here is a doc. Actually, this rule has been defined in the database_global.json also: https://github.com/sonic-net/sonic-buildimage/blob/ada7c6a72e27a9729628f383f48cd5f75d4da227/dockers/docker-database/database_global.json.j2#L30

Reply:
Done. Update a paragraph to describe the database_global.json.

Reply:
var/run/redisdpu{DPU_ID} -> var/run/redisdpu{database_name}.
Maybe you should remove this part, because it is part of database_global.json